blssig/src/verifier/mod.rs

```diff
@@ -0,0 +1,168 @@
+
+use crate::bdn::BDNAggregation;
+
+#[cfg(test)]
+mod tests;
+
+#[derive(Error, Debug, PartialEq)]
+pub enum BLSError {
+    #[error("empty public keys provided")]
+    EmptyPublicKeys,
+    #[error("invalid public key length: expected 48 bytes, got {0}")]
+    InvalidPublicKeyLength(usize),
+    #[error("failed to deserialize public key")]
+    PublicKeyDeserialization,
+    #[error("invalid signature length: expected 96 bytes, got {0}")]
+    InvalidSignatureLength(usize),
+    #[error("failed to deserialize signature")]
+    SignatureDeserialization,
+    #[error("BLS signature verification failed")]
+    SignatureVerificationFailed,
+    #[error("mismatched number of public keys and signatures: {pub_keys} != {sigs}")]
+    LengthMismatch { pub_keys: usize, sigs: usize },
+}
+
+/// BLS signature verifier using BDN aggregation scheme
+///
+/// This verifier implements the same scheme used by go-f3/blssig, with:
+/// - BLS12_381 curve
+/// - G1 for public keys, G2 for signatures  
+/// - BDN aggregation for rogue-key attack prevention
+/// - Point cache for performance optimization
+pub struct BLSVerifier {
+    /// Cache for deserialized public key points to avoid expensive repeated operations
+    point_cache: RwLock<HashMap<Vec<u8>, PublicKey>>,
+}
+
+impl Default for BLSVerifier {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Maximum number of cached public key points to prevent excessive memory usage
+const MAX_POINT_CACHE_SIZE: usize = 10_000;
+
+impl BLSVerifier {
+    pub fn new() -> Self {
+        Self {
+            point_cache: RwLock::new(HashMap::new()),
+        }
+    }
+
+    /// Verifies a single BLS signature
+    fn verify_single(&self, pub_key: &PubKey, msg: &[u8], sig: &[u8]) -> Result<(), BLSError> {
+        // Validate input lengths
+        if pub_key.0.len() != 48 {
+            return Err(BLSError::InvalidPublicKeyLength(pub_key.0.len()));
+        }
+        if sig.len() != 96 {
+            return Err(BLSError::InvalidSignatureLength(sig.len()));
+        }
+
+        // Get cached public key
+        let pub_key = self.get_or_cache_public_key(&pub_key.0)?;
+
+        // Deserialize signature
+        let signature =
+            Signature::from_bytes(sig).map_err(|_| BLSError::SignatureDeserialization)?;
+
+        // Verify using bls-signatures
+        let messages = vec![msg];
+        let public_keys = vec![pub_key];
+        match verify_messages(&signature, &messages, &public_keys) {
+            true => Ok(()),
+            false => Err(BLSError::SignatureVerificationFailed),
+        }
+    }
+
+    /// Gets a cached public key or deserializes and caches it
+    fn get_or_cache_public_key(&self, pub_key_bytes: &[u8]) -> Result<PublicKey, BLSError> {
+        // Check cache first
+        {
+            let cache = self.point_cache.read();
+            if let Some(cached_key) = cache.get(pub_key_bytes) {
+                return Ok(*cached_key);
+            }
+        }
```

Comment:
nit(code golf): one less line if you `drop(cache)` instead of making scopes.

Reply:
Fixed here: https://github.com/ChainSafe/rust-f3/pull/27/commits/7a0a39befd832eeb7601cd2aa66c95b2d5d29c4a